diesel/src/sqlite/connection/stmt.rs

```diff
@@ -126,18 +220,159 @@ impl Drop for Statement {
     }
 }
 
+// A warning for future editiors:
+// Changing this code to something "simplier" may
+// introduce undefined behaviour. Make sure you read
+// the following discussions for details about
+// the current version:
+//
+// * https://github.com/weiznich/diesel/pull/7
+// * https://users.rust-lang.org/t/code-review-for-unsafe-code-in-diesel/66798/
+// * https://github.com/rust-lang/unsafe-code-guidelines/issues/194
+struct BoundStatement<'stmt, 'query> {
+    statement: MaybeCached<'stmt, Statement>,
+    // we need to store the query here to ensure noone does
+    // drop it till the end ot the statement
+    // We use a boxed queryfragment here just to erase the
+    // generic type, we use NonNull to communicate
+    // that this is a shared buffer
+    query: Option<NonNull<dyn QueryFragment<Sqlite> + 'query>>,
+    // we need to store any owned bind values speratly, as they are not
+    // contained in the query itself. We use NonNull to
+    // communicate that this is a shared buffer
+    binds_to_free: Vec<(i32, Option<NonNull<[u8]>>)>,
+}
+
+impl<'stmt, 'query> BoundStatement<'stmt, 'query> {
+    fn bind<T>(
+        statement: MaybeCached<'stmt, Statement>,
+        query: T,
+    ) -> QueryResult<BoundStatement<'stmt, 'query>>
+    where
+        T: QueryFragment<Sqlite> + QueryId + 'query,
+    {
+        // Don't use a trait object here to prevent using a virtual function call
+        // For sqlite this can introduce a measurable overhead
+        let query = Box::new(query);
+
+        let mut bind_collector = SqliteBindCollector::new();
+        query.collect_binds(&mut bind_collector, &mut ())?;
+        let SqliteBindCollector { binds } = bind_collector;
+
+        let mut ret = BoundStatement {
+            statement,
+            query: None,
+            binds_to_free: Vec::with_capacity(
+                binds
+                    .iter()
+                    .filter(|&(b, _)| {
+                        matches!(
+                            b,
+                            SqliteBindValue::BorrowedBinary(_)
+                                | SqliteBindValue::BorrowedString(_)
+                                | SqliteBindValue::String(_)
+                                | SqliteBindValue::Binary(_)
+                        )
+                    })
+                    .count(),
```

Comment:
This should be done as a `Vec::reserve` in `bind_buffers` so that code is not duplicated in every monomorphization of `bind`.

Reply:
Also we may want to specify why we do this as it's not obvious form my POV:
```rust
            // It is useful to preallocate `binds_to_free` because it
            // - Guarantees that pushing inside it cannot panic, which guarantees the `Drop`
            //   impl of `BoundStatement` will always re-`bind` as needed
            // - Avoids reallocations
```